sql/wsrep_mysqld.cc

```diff
@@ -3664,3 +3664,107 @@ bool wsrep_table_list_has_non_temp_tables(THD *thd, TABLE_LIST *tables)
   }
   return false;
 }
+
+bool wsrep_foreign_key_append(THD *thd)
+{
+  List <FOREIGN_KEY_INFO> fk_list;
+  FOREIGN_KEY_INFO *fk;
+  bool error= false;
+  int rcode= 0;
+
+  if (WSREP(thd) && !thd->wsrep_applier &&
```

Comment:
Here we access THD internals, my question is it possible to do KILL QUERY|CONNECTION while we are adding these foreign keys leading to race condition or crash. To avoid this we could take THD::LOCK_thd_data mutex, but better to do a test case first, i.e. add DEBUG_SYNC below when we start accessing fk list and from other connection try to KILL this query.

Reply:
Mutex protection is necessary only if the internal accessed THD state is shared between the owner thread and the thread which executes KILL query. This does not seem to be the case here (see `THD::awake_no_mutex()` what state is accessed by the KILL.

IMO it does not make sense to make a KILL related test for every possible scenario where THD internals are accessed as THD internals are involved in all query processing.